src/main/java/com/epam/msu/service/impl/CertificateServiceImpl.java

```diff
@@ -0,0 +1,125 @@
+package com.epam.msu.service.impl;
+
+import com.epam.msu.dao.impl.CertificateDaoImpl;
+import com.epam.msu.dao.impl.TagDaoImpl;
+import com.epam.msu.dto.CertificateDto;
+import com.epam.msu.entity.Certificate;
+import com.epam.msu.entity.Tag;
+import com.epam.msu.service.CertificateService;
+import com.epam.msu.util.MappingUtils;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.util.ArrayList;
+import java.util.List;
+
+@Service
+public class CertificateServiceImpl implements CertificateService {
+
+    private final CertificateDaoImpl certificateDao;
+    private final TagDaoImpl tagDao;
+
+    @Autowired
+    public CertificateServiceImpl(CertificateDaoImpl certificateDao, TagDaoImpl tagDao) {
+        this.certificateDao = certificateDao;
+        this.tagDao = tagDao;
+    }
+
+    @Override
+    public List<CertificateDto> getAllCertificates() {
+        List<Certificate> certificateList = certificateDao.getAllCertificates();
+        List<CertificateDto> certificatesDto = MappingUtils.mapToListCertificateDto(certificateList);
+        List<Tag> tags = new ArrayList<>();
+        for (CertificateDto certificate : certificatesDto) {
+            List<Tag> tag = tagDao.getTagsByCertificateId((int) certificate.getId());
+            certificate.setTag(tag);
+
+        }
+        ;
+        return certificatesDto;
+    }
+
+    @Override
+    public CertificateDto getCertificateById(int id) {
+        Certificate certificate = certificateDao.getCertificateById(id);
+        CertificateDto certificateDto = null;
+        if (certificate != null) {
+            certificateDto = MappingUtils.mapToCertificateDto(certificate);
+            List<Tag> certificateTag = tagDao.getTagsByCertificateId(id);
+            certificateDto.setTag(certificateTag);
+            return certificateDto;
+        }
+        return null;
+    }
+
+    @Override
+    public CertificateDto createNewCertificate(CertificateDto certificate) {
+        List<Tag> tags = certificate.getTag();
+        int tagId = 0;
+        certificate = MappingUtils.mapToCertificateDto(certificateDao.createNewCertificate(MappingUtils.mapToCertificate(certificate)));
+        int certificateId = certificateId = (int) certificate.getId();
+        for (Tag tag : tags) {
+            if (isUniqueTag(tag)) {
+                tag = tagDao.createTag(tag);
+                tagId = (int) tag.getId();
+                certificateDao.addInIntermediateTable(certificateId, tagId);
+            } else {
+                String tagName = tag.getName();
+                Tag existsTag = tagDao.getTagByName(tagName);
+                tagId = (int) existsTag.getId();
+                certificateDao.addInIntermediateTable(certificateId, tagId);
+            }
+        }
+        certificate.setTag(tags);
+        return certificate;
+    }
+
+    private boolean isUniqueTag(Tag tag) {
+        List<Tag> tags = tagDao.getAllTags();
```

Comment:
A little bit strange logic. Each time this method is invoked, all tags are retrieved from database(when there are 100-200 tags t doesn't influence on performance, but in case of a lot of tags - it will impact). It is better to create separate method in dao existsByName, for example, and check tags one by one

Reply:
Fixed